File: backend/internal/models/session.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type SessionStatus string

const (
	SessionPending    SessionStatus = "pending"
	SessionPaid       SessionStatus = "paid"
	SessionCharging   SessionStatus = "charging"
	SessionCompleted  SessionStatus = "completed"
	SessionCancelled  SessionStatus = "cancelled"
	SessionFailed     SessionStatus = "failed"
)

type ChargingSession struct {
	ID           uuid.UUID      `gorm:"type:uuid;primaryKey" json:"id"`
	UserID       uuid.UUID      `gorm:"type:uuid;not null;index" json:"user_id"`
	ConnectorID  uuid.UUID      `gorm:"type:uuid;not null;index" json:"connector_id"`
	Status       SessionStatus  `gorm:"size:20;default:'pending'" json:"status"`
	EnergyKWH    float64        `gorm:"type:decimal(10,3);default:0" json:"energy_kwh"`
	PowerKW      float64        `gorm:"type:decimal(6,2);default:0" json:"power_kw"`
	Progress     int            `gorm:"default:0" json:"progress"`
	TotalCost    float64        `gorm:"type:decimal(12,2);default:0" json:"total_cost"`
	TargetKWH    float64        `gorm:"type:decimal(10,3);default:0" json:"target_kwh"`
	StartedAt    *time.Time     `json:"started_at"`
	EndedAt      *time.Time     `json:"ended_at"`
	CreatedAt    time.Time      `json:"created_at"`
	UpdatedAt    time.Time      `json:"updated_at"`
	DeletedAt    gorm.DeletedAt `gorm:"index" json:"-"`

	User      User      `gorm:"foreignKey:UserID" json:"user,omitempty"`
	Connector Connector `gorm:"foreignKey:ConnectorID" json:"connector,omitempty"`
	Payment   *Payment  `gorm:"foreignKey:SessionID" json:"payment,omitempty"`
}

func (s *ChargingSession) BeforeCreate(tx *gorm.DB) error {
	if s.ID == uuid.Nil {
		s.ID = uuid.New()
	}
	return nil
}

File: mobile/lib/models/session.dart
import 'station.dart';

class ChargingSession {
  final String id;
  final String userId;
  final String connectorId;
  final String status;
  final double energyKWH;
  final double powerKW;
  final int progress;
  final double totalCost;
  final double targetKWH;
  final DateTime? startedAt;
  final DateTime? endedAt;
  final DateTime createdAt;
  final Connector? connector;

  ChargingSession({
    required this.id,
    required this.userId,
    required this.connectorId,
    required this.status,
    required this.energyKWH,
    required this.powerKW,
    required this.progress,
    required this.totalCost,
    required this.targetKWH,
    this.startedAt,
    this.endedAt,
    required this.createdAt,
    this.connector,
  });

  factory ChargingSession.fromJson(Map<String, dynamic> json) {
    return ChargingSession(
      id: json['id'] ?? '',
      userId: json['user_id'] ?? '',
      connectorId: json['connector_id'] ?? '',
      status: json['status'] ?? '',
      energyKWH: (json['energy_kwh'] ?? 0).toDouble(),
      powerKW: (json['power_kw'] ?? 0).toDouble(),
      progress: json['progress'] ?? 0,
      totalCost: (json['total_cost'] ?? 0).toDouble(),
      targetKWH: (json['target_kwh'] ?? 0).toDouble(),
      startedAt: json['started_at'] != null
          ? DateTime.tryParse(json['started_at'])
          : null,
      endedAt: json['ended_at'] != null
          ? DateTime.tryParse(json['ended_at'])
          : null,
      createdAt: DateTime.tryParse(json['created_at'] ?? '') ?? DateTime.now(),
      connector: json['connector'] != null
          ? Connector.fromJson(json['connector'])
          : null,
    );
  }

  bool get isActive =>
      status == 'pending' || status == 'paid' || status == 'charging';

  bool get isCharging => status == 'charging';

  bool get isCompleted => status == 'completed';

  String get statusLabel {
    switch (status) {
      case 'pending':
        return 'Menunggu Pembayaran';
      case 'paid':
        return 'Dibayar';
      case 'charging':
        return 'Sedang Charging';
      case 'completed':
        return 'Selesai';
      case 'cancelled':
        return 'Dibatalkan';
      case 'failed':
        return 'Gagal';
      default:
        return status;
    }
  }
}
